platform/src/main/java/org/stellar/anchor/platform/callback/RestRateIntegration.java

```diff
@@ -68,29 +83,131 @@ public GetRateResponse getRate(GetRateRequest request) throws AnchorException {
     try {
       getRateResponse = gson.fromJson(responseContent, GetRateResponse.class);
     } catch (Exception e) { // cannot read body from response
-      Log.errorEx("Error parsing body response to GetRateResponse", e);
+      errorEx("Error parsing body response to GetRateResponse", e);
       throw new ServerErrorException("internal server error", e);
     }
 
+    validateRateResponse(request, getRateResponse);
+    return getRateResponse;
+  }
+
+  void validateRateResponse(GetRateRequest request, GetRateResponse getRateResponse)
+      throws ServerErrorException {
+    AssetInfo sellAsset = assetService.getAssetByName(request.getSellAsset());
     GetRateResponse.Rate rate = getRateResponse.getRate();
     if (rate == null || rate.getPrice() == null) {
-      Log.error("missing 'price' in the GET /rate response");
-      throw new ServerErrorException("internal server error");
+      logErrorAndThrow("missing 'price' in the GET /rate response", ServerErrorException.class);
     }
 
     if (request.getType() != GetRateRequest.Type.INDICATIVE) {
-      if (rate.getFee() == null) {
-        Log.error("'fee' is missing in the GET /rate response");
-        throw new ServerErrorException("internal server error");
+      if (Objects.requireNonNull(rate).getFee() == null) {
+        logErrorAndThrow("'fee' is missing in the GET /rate response", ServerErrorException.class);
       }
     }
 
     if (request.getType() == GetRateRequest.Type.FIRM) {
       if (rate.getId() == null || rate.getExpiresAt() == null) {
-        Log.error("'id' and/or 'expires_at' are missing in the GET /rate response");
-        throw new ServerErrorException("internal server error");
+        logErrorAndThrow(
+            "'id' and/or 'expires_at' are missing in the GET /rate response",
+            ServerErrorException.class);
+      }
+    }
+
+    // sell_amount is present and positive number
+    if (!isPositiveNumber(Objects.requireNonNull(rate).getSellAmount())) {
```

Comment:
We should also validate that sell, buy and fee amounts have the scale corresponding to their asset's scale

Reply:
Good suggestions. Tests added.